### PR TITLE
feat: passkey provider

### PR DESCRIPTION
**Pull request by antoncoding**
## Features
- [x] `useAccount` ( => `usePasskeyAccount`) to expose wallet address
- [x] login, register
- [x] cache logged in user 
- [x] smoothier reconnect state (animation ...etc)
- [x] logout
- [x] sendTransactions!

## TLDR
* 使用 `PasskeyProvider`, 用 `usePasskeyAccount` expose address
* 當網頁開啟時，自動使用 `reconnect` 嘗試用上次得到的 passkey 資料來創建 passkey
* logout 會清除儲存的 passkey 資料，之後便不會重新連線
* 送出交易使用 sendTransactions，仍然使用 wagmi hook 讀取交易 receipt 狀態

## Impact
* Strava 登入玩不用重新按 Login ！
* 每次開啟 app 不需要案 Login ！

### CODE REVIEW

File: web/src/providers/PasskeyProvider.tsx [hunk 1]
```diff
@@ -0,0 +1,235 @@
+'use client';
+
+import React, { createContext, useContext, useState, useEffect, useCallback, useMemo } from 'react';
```
Comment (ryanycw) [inline web/src/providers/PasskeyProvider.tsx:3]:
```suggestion
import { createContext, useContext, useState, useEffect, useCallback, useMemo, ReactNode } from 'react';
```

File: web/src/providers/PasskeyProvider.tsx [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+  accountClient: undefined,
+  login: async () => {},
+  register: async () => {},
+  logout: () => {},
+};
+
+// Use the default context value when creating the context
+const PasskeyContext = createContext<PasskeyContextType>(defaultContextValue);
+
+export function PasskeyProvider({ children }: { children: React.ReactNode }) {
```
Comment (ryanycw) [inline web/src/providers/PasskeyProvider.tsx:66]:
```suggestion
export function PasskeyProvider({ children }: { children: ReactNode }) {
```

File: web/src/providers/PasskeyProvider.tsx [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+  const [address, setAddress] = useState<Hex | undefined>(undefined);
```
Comment (ryanycw) [inline web/src/providers/PasskeyProvider.tsx:67]: Can consider to have a PasskeyState to manage all the related state here.

**Review by Ankarrr — APPROVED** [review]
測試過 preview，主要 review passkey provider，看起來 ok

終於不用一直登入了~

File: web/src/hooks/usePasskeyConnection.ts [hunk 4]
```diff
@@ -1,35 +1,46 @@
...
 import { logEventSimple } from '@/utils/gtag';
+import { usePasskeyAccount } from '@/providers/PasskeyProvider';
 import storage from 'local-storage-fallback';
-import { useEffect, useState } from 'react';
-import { useConnect } from 'wagmi';
+import { useEffect, useState, useCallback } from 'react';
 
 export default function usePasskeyConnection() {
-  const { connectors, connect, isPending } = useConnect();
   const [initializing, setInitializing] = useState(true);
```
Comment (ryanycw) [inline web/src/hooks/usePasskeyConnection.ts:9]: discussions: do we need this state? Wouldn't this be the same value as isInitializing?

File: web/src/hooks/transaction/useSubmitTransaction.ts [hunk 5]
```diff
@@ -1,18 +1,19 @@
...
+  Abi,
+  DecodeEventLogReturnType,
+  TransactionReceipt,
+  decodeEventLog,
+  encodeFunctionData,
+  http,
+} from 'viem';
+import { usePasskeyAccount } from '@/providers/PasskeyProvider';
+import { useWaitForTransactionReceipt } from 'wagmi';
+import { getChainsForEnvironment } from '@/store/supportedChains';
```
Comment (ryanycw) [inline web/src/hooks/transaction/useSubmitTransaction.ts:13]:
```suggestion
```

File: web/src/hooks/transaction/useSubmitTransaction.ts [hunk 6]
```diff
@@ -1,18 +1,19 @@
...
-import { useWriteContracts, useCallsStatus } from 'wagmi/experimental';
-
-const zerodevApiKey = process.env.NEXT_PUBLIC_ZERODEV_API_KEY;
-const ZERODEV_PAYMASTER_URL = `https://rpc.zerodev.app/api/v3/paymaster/${zerodevApiKey}`;
+  Abi,
+  DecodeEventLogReturnType,
+  TransactionReceipt,
+  decodeEventLog,
+  encodeFunctionData,
+  http,
```
Comment (ryanycw) [inline web/src/hooks/transaction/useSubmitTransaction.ts:9]:
```suggestion
```

File: web/src/hooks/transaction/useSubmitTransaction.ts [hunk 7]
```diff
@@ -1,18 +1,19 @@
-import { useEffect, useState } from 'react';
+import { useEffect, useMemo, useState, useRef } from 'react';
```
Comment (ryanycw) [inline web/src/hooks/transaction/useSubmitTransaction.ts:1]:
```suggestion
import { useEffect, useState, useRef } from 'react';
```

File: web/src/hooks/transaction/useSubmitTransaction.ts [hunk 8]
```diff
@@ -59,104 +43,89 @@ const useSubmitTransaction = (
...
-    data: batchHash,
-    error: batchWriteContractError,
-    isError: isBatchWriteContractError,
-    isPending: isBatchWriteContractLoading,
-    isSuccess: isBatchWriteContractSuccess,
+  const { accountClient, account } = usePasskeyAccount();
 
-    reset: resetBatchWriteContract,
-  } = useWriteContracts();
+  console.log('tx error', error);
```
Comment (ryanycw) [inline web/src/hooks/transaction/useSubmitTransaction.ts:52]:
```suggestion
```

File: web/src/hooks/transaction/useSubmitTransaction.ts [hunk 9]
```diff
@@ -1,18 +1,19 @@
...
+  encodeFunctionData,
+  http,
+} from 'viem';
+import { usePasskeyAccount } from '@/providers/PasskeyProvider';
+import { useWaitForTransactionReceipt } from 'wagmi';
+import { getChainsForEnvironment } from '@/store/supportedChains';
 
 const getEvents = (
-  contractCallConfig: UseSimulateContractParameters,
+  contractCallConfig: any,
```
Comment (ryanycw) [inline web/src/hooks/transaction/useSubmitTransaction.ts:16]: Will it cause any issue to remain using UseSimulateContractParameters for type safety?

File: web/src/hooks/transaction/useSubmitTransaction.ts [hunk 10]
```diff
@@ -59,104 +43,89 @@ const useSubmitTransaction = (
...
-  } = useWaitForTransactionReceipt({
-    hash: callsStatus?.receipts?.[0]?.transactionHash as `0x${string}`,
-  });
+  const { data: transactionReceipt, isLoading: isWaitForTransactionLoading } =
+    useWaitForTransactionReceipt({
+      hash: transactionHash,
+    });
 
-  const rawError = waitForTransactionError ?? batchWriteContractError ?? simulateContractError;
+  const submitTransaction = () => {
```
Comment (ryanycw) [inline web/src/hooks/transaction/useSubmitTransaction.ts:59]: I think we can use `useCallback` to optimize it

File: web/src/hooks/transaction/useSubmitTransaction.ts [hunk 11]
```diff
@@ -59,104 +43,89 @@ const useSubmitTransaction = (
...
+  }, [transactionReceipt, options, contractCallConfig]);
 
-    if (error) {
-      onError?.(error, rawError);
-      setIsLoading(false);
-      resetBatchWriteContract();
-    }
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [transactionReceipt, isSuccess, isError]);
+  useEffect(() => {
```
Comment (ryanycw) [inline web/src/hooks/transaction/useSubmitTransaction.ts:114]: In my understanding, we won't need this part. Cause if the transaction failed, this hook will be triggered and create a new mount anyway.

**Review by ryanycw — APPROVED** [review]

File: web/src/providers/PasskeyProvider.tsx (hunk 3 above)
Comment (antoncoding) [reply web/src/providers/PasskeyProvider.tsx:67]: Claude said this:

Given your current setup and the relatively small number of state variables, I believe the current approach with separate states is more appropriate and maintainable. It provides a good balance between clarity, flexibility, and ease of use. If your application grows significantly in complexity, you can always refactor to a single state object or even consider using a more robust state management solution like Redux or MobX.

**Pull request merged**